### ISSUE TITLE
How to Create a Custom Authentication Provider for OpenID Connect Server

### ISSUE THREAD

**Issue by jayonline9** [mitreid-connect/OpenID-Connect-Java-Spring-Server#774]
I would like to use my own authentication mechanism (preferably Biometrics) for user authentication. Can you provide some advice on how to go about it???

**Comment by jricher** [comment]
MITREid Connect uses Spring Security for its authentication, so you can put whatever component you like into that space. There are lots of good resources on the web about how to write and configure Spring Security filters for custom authentication mechanisms. 

You'll want to look at the `user-context.xml` file for where the user authentication is defined. In the core project this is a simple username/password field against a local database. In others like the LDAP overlay project, this connects to an LDAP server. In some systems, like MIT's "oidc.mit.edu" server, there are actually a handful of different authentication mechanisms that can be used in parallel: LDAP, kerberos, and certificates in that case. 

Note that in all cases, you'll still need to have access to a `UserInfo` data store somewhere. This can be sourced from the database, from LDAP, or from something else, but it needs to be available for each logged in user.

**Comment by jayonline9** [comment]
Thanks for replying Justin...I wrote a custom authentication provider which checks for username and password and was able to successfully login to the OpenID connect server. 
But my problem is, once I authorize the Simple Web Application, I am getting a 'HTTP Status 401 - Authentication Failed: Token Endpoint did not return an id_token' from Tomcat. Can you please advice on how to resolve this issue.
Should I raise this a new issue?

**Comment by jricher** [comment]
You might not be asking for the `openid` scope, which is what turns on the id token. Also make sure the userinfo is present as described above. You can confirm this by logging in to your server and checking for profile information.
